src/CodeGen_ARM.cpp

```diff
@@ -328,15 +521,18 @@ void CodeGen_ARM::visit(const Cast *op) {
                         // The arm32 llvm backend wants right shifts to come in as negative values.
                         shift_amount = -shift_amount;
                     }
-                    Value *shift = nullptr;
-                    // The arm64 llvm backend wants i32 constants for right shifts.
-                    if (target.bits == 64 && pattern.type == Pattern::RightShift) {
-                        shift = ConstantInt::get(i32_t, shift_amount);
+                    // TODO: It would be nice if call_overloaded_intrin could handle this type promotion.
+                    Expr b;
+                    if (matches[1].type().is_scalar()) {
+                        if (target.bits == 32) {
+                            b = make_const(matches[0].type().with_code(halide_type_int), shift_amount);
+                        } else {
+                            b = make_const(Int(32), shift_amount);
```

Comment:
Would be good to preserve the comment that the arm64 backend wants the type to always be int32